type이 LOGIN 일시 accessToken, refreshToken 발급

sin-hyunjin opened the pull request:
## Issues 번호 :

Closes #17 

## 변경, 추가된 코드(설명 등)



![image](https://github.com/Chukapoka/server/assets/116487398/c34cfece-ad40-42a9-9064-b927d2fde599)
### - [ Accese_Token & Refresh_Token 블로그 참고(1)](https://9keyyyy.tistory.com/48)
### - [ Accese_Token & Refresh_Token 블로그 참고(2)](https://velog.io/@songunnie/Project-Spring-Security-JWT-Refresh-Token-%EC%82%AC%EC%9A%A9%ED%95%B4%EB%B3%B4%EA%B8%B0)
- type이 LOGIN 일시 accessToken, refreshToken발급 확인 


## 코드 주의점

-
-
-
-

<!-- This is an auto-generated comment: release notes by OSS CodeRabbit -->
### Summary by CodeRabbit

**릴리스 노트: 2024-01-25**

변경 사항은 다음과 같이 분류됩니다:

- New Feature:
    - `JwtAuthenticationFilter` 클래스에 `resolveToken` 메서드가 추가되었습니다. 이 메서드는 HTTP 요청의 헤더에서 토큰을 추출하는 역할을 합니다.
    - 새로운 클래스 `JwtTokenProvider`가 도입되었습니다. 이 클래스는 JWT 토큰을 생성하고 유효성을 검사하는 역할을 담당합니다. 사용자 인증을 기반으로 토큰을 생성하고, 토큰에서 사용자 정보를 추출하며, 토큰의 정당성을 검증하는 메서드가 포함되어 있습니다. 이 클래스는 JWT 작업을 위해 `io.jsonwebtoken` 라이브러리를 사용합니다.

- Refactor:
    - Spring Security 구성 클래스인 `SecurityConfig`에 대한 변경 사항입니다. `JwtAuthenticationFilter`가 `UsernamePasswordAuthenticationFilter` 앞에 추가되었습니다. 이 변경은 코드의 외부 인터페이스나 동작에 영향을 미치지 않으며, 기능적인 수정도 없습니다.

- Documentation:
    - `BaseResponse` 클래스에 새로운 생성자가 추가되었습니다. 이 생성자는 `ResultType`, `UserResponseDto`, `TokenDto`를 매개변수로 받습니다. 그러나 현재 구현은 빈 메서드로 남겨져 있습니다.
    - `CustomUser` 클래스에 `userId` 필드가 추가되었습니다. 이 필드는 생성자에서 초기화되며, `getUserId()` 메서드를 통해 접근할 수 있습니다.

- Style:
    - `TokenDto` 클래스에 `grantType` 필드가 추가되었습니다. 이 필드는 JWT 인증 타입을 나타내며, HTTP 헤더에 prefix로 붙여주는 역할을 합니다.

- Bug fix:
    - `Authority` 열거형에 `ROLE_ADMIN`이라는 새로운 값이 추가되었습니다. 또한 `getByAuthority`라는 정적 메서드가 추가되어 authority 문자열에 해당하는 `Authority` 열거형 값을 검색하는 기능을 제공합니다.

- Chore:
    - `UserController`의 `authenticateUser` 함수에서 변수 이름이 변경되었습니다.

각 변경 사항은 해당하는 분류에 따라 요약되었습니다. 이 릴리스 노트는 코드 변경 사항을 간결하게 설명하고, 사용자에게 전달될 수 있는 주요 업데이트를 강조합니다.
<!-- end of auto-generated comment: release notes by OSS CodeRabbit -->

sin-hyunjin committed:
type이 LOGIN 일시 accessToken, refreshToken 발급

sin-hyunjin committed:
RefreshToken 기한 만료시 재발급 로직 진행중

sin-hyunjin committed:
RefreshToken 기한 만료시 재발급 확인완료

sin-hyunjin committed:
회원가입시 ROLE_USER 권한추가

sin-hyunjin commented:
## 로그아웃 :  Access Token, Refresh Token  확인후 사용자 로그아웃 진행

![스크린샷 2024-01-29 오후 9 23 38](https://github.com/Chukapoka/server/assets/116487398/bcd54ff4-4035-4690-8a5a-02ecab8e77d3)

<hr>

## 로그아웃 된 사용자일 경우 Exception 처리

![image](https://github.com/Chukapoka/server/assets/116487398/f119f5a6-0b6f-47c6-9422-848c13bf63a8)

sin-hyunjin committed:
로그아웃 : 유저의 토큰 값 확인 후 저장소에서 토큰 제거

sin-hyunjin committed:
로그아웃 : response 값을 성공, 실패로만 분류

sin-hyunjin committed:
tb_token 저장소에 accese token 저장 추가(refreshToken관련된 파일이름 -> Token으로 파일이름변경)

doyou1 commented on src/main/java/com/chukapoka/server/common/dto/TokenDto.java:
```diff
@@ -0,0 +1,20 @@
+import lombok.Data;
+/**
+ * 클라이언트에 토큰을 보내기 위한 DTO
+ * @author rimsong
```
지우셈!

sin-hyunjin replied:
삭제완료

doyou1 commented on src/main/resources/application.yaml:
```diff
@@ -1,6 +1,6 @@
+#   active: dev
 #    active: dev-db
-#     active: local
+     active: local
```
깃허브에 올라오는 application.yaml은 항상 `active': dev`

sin-hyunjin replied:
커밋할 때 주의할게욤

sin-hyunjin committed:
application local -> dev 변경

doyou1 commented on src/main/java/com/chukapoka/server/WebConfig.java:
```diff
@@ -1,6 +1,7 @@
 package com.chukapoka.server;
 
 import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.context.annotation.Bean;
```
사용하지 않는 import문 삭제

sin-hyunjin replied:
WebConfig :  import org.springframework.context.annotation.Bean;
UserRepository : import com.chukapoka.server.common.enums.EmailType 
User : import org.springframework.security.core.authority.SimpleGrantedAuthority; , import java.util.List; 
UserResponDto : import com.chukapoka.server.user.entity.User;
UserRequestDto : import jakarta.validation.constraints.NotNull;
EmailCheckResponseDto : import com.chukapoka.server.common.enums.NextActionType;
BaseResponse : import com.chukapoka.server.user.dto.UserResponseDto;
JwtTokenProvider : import com.chukapoka.server.user.entity.User;

필요없는 import문 삭제완료

doyou1 commented on src/main/java/com/chukapoka/server/common/authority/JwtTokenProvider.java:
```diff
@@ -0,0 +1,153 @@
+                .map(GrantedAuthority::getAuthority)
+                .collect(Collectors.joining(","));
+
+        Date now = new Date();
+        Date accessTokenExpiresIn = new Date(now.getTime() + ACCESS_EXPIRATION_MILLISECONDS);
+        Date refreshExpiration = new Date(now.getTime() + REFRESH_EXPIRATION_MILLISECONDS);
```
Good 👍

doyou1 commented on src/main/java/com/chukapoka/server/common/dto/BaseResponse.java:
```diff
@@ -2,6 +2,7 @@
 
 
 import com.chukapoka.server.common.enums.ResultType;
+import com.chukapoka.server.user.dto.UserResponseDto;
```
불필요한 import? 아니면 BaseResponse 항목추가?

sin-hyunjin replied:
BaseResponse 테스트 해봤는데 import삭제 안되어서 삭제 완료

doyou1 commented on src/main/java/com/chukapoka/server/common/entity/Token.java:
```diff
@@ -0,0 +1,41 @@
+
+    @Id
+    @Column(name = "token_key")
+    private String key;
```
현재는 email로 되어있지만, email은 소셜로그인 등으로 중복될 가능성이 있기에 `userId`로 변경한다. 
변경하는 방법은 

```
@Getter
public class CustomUser extends User {
    private final String email;

    public CustomUser(String userId, String email, String password, Collection<? extends GrantedAuthority> authorities) {
        super(userId, password, authorities);
        this.email = email;
    }

}
```

sin-hyunjin replied:
![image](https://github.com/Chukapoka/server/assets/116487398/649b8784-ea36-4d09-b989-f5c1da43648d)
코드수정후  email값이 아닌 id값으로 넘어오는것 확인

doyou1 commented on src/main/java/com/chukapoka/server/common/repository/TokenRepository.java:
```diff
@@ -0,0 +1,11 @@
+
+import java.util.Optional;
+
+public interface TokenRepository extends JpaRepository<Token,Long> {
```
확인 부탁

sin-hyunjin replied:
```java 
public interface TokenRepository extends JpaRepository<Token,String> {
    Optional<Token> findByKey(String key);
}
```

Long -> String으로 변경

sin-hyunjin replied:
CustomUserId 값이 Long값이여서 다시 Long으로 수정

doyou1 commented on src/main/java/com/chukapoka/server/common/authority/SecurityConfig.java:
```diff
@@ -11,22 +11,35 @@
                 .csrf(AbstractHttpConfigurer::disable)
+                .addFilterBefore(new JwtAuthenticationFilter(jwtTokenProvider), UsernamePasswordAuthenticationFilter.class)
                 .authorizeHttpRequests((authorizeRequests) -> {
                     authorizeRequests
                             .requestMatchers("/api/user/emailcheck", "/api/user/**").anonymous();
```
authorizeRequests
                            .requestMatchers("/api/user/emailcheck", "/api/user", "/api/user/authNumber").anonymous()
                            // hasRole("MEMBER") = 유효한 토큰을 header에 집어넣은 채 request를 했다.
                            .requestMatchers("/api/member/**").hasRole("MEMBER")
                            .anyRequest().permitAll();

sin-hyunjin replied:
SecurityConfig 회원 권한 수정완료

doyou1 reviewed (changes requested):
good

doyou1 commented:
@sin-hyunjin 
토큰은 항상 header에

sin-hyunjin committed:
application local -> dev 변경

sin-hyunjin committed:
CustomUser UserId값으로 가져오기
controller에서 acceseToken값 확인해보기

sin-hyunjin committed:
api/user/test에 Headers에 accessToken값을 넣을시 권한설정 및 반환값 확인

sin-hyunjin committed:
@RequestHeader 로 로그아웃 실행시 accessToken값으로만 확인 후 로그아웃

sin-hyunjin committed:
회원 권한 접근 성공
- CustomUserDetailsService 추가

sin-hyunjin committed:
회원 권한이 있는 userId값과 TOKENDB에 있는 Id 값이 같다면 로그아웃 되면서 refreshKoken 삭제 진행

sin-hyunjin committed:
로그아웃 req 삭제

sin-hyunjin committed:
Token Valid 값 추가

doyou1 commented on src/main/java/com/chukapoka/server/common/authority/JwtAuthenticationFilter.java:
```diff
@@ -0,0 +1,51 @@
+    public void doFilter(ServletRequest request, ServletResponse response, FilterChain chain) throws IOException, ServletException {
+        // 1. Request Header 에서 토큰을 꺼냄
+        String token = resolveToken((HttpServletRequest) request);
+        System.out.println("Received token: " + token);
```
`println`은 지우고 올릴것!

sin-hyunjin replied:
확인이요!

doyou1 commented on README.md:
```diff
@@ -41,4 +41,16 @@
\ No newline at end of file
+  - CorsConfig:  서로 다른 Server 환경에서 자원을 공유에 필요한 설정
+
+## 인가 표현식 
+
+![API.png](screenshots%2FAPI.png)
+
+hasRole과 hasAuthority 차이
+
+hasRole은 괄호안에 USER라 넣으면 ROLEUSER로 변환
+[그러니까 ROLE PREFIX를 넣지 않아야한다]
+hasAuthority는 ROLE_USER로 PREFIX를 넣어야한다.
+
+hasAnyRole,Authority는 여러 권한을 넣는 것.
```
Good 🚀

doyou1 commented on src/main/java/com/chukapoka/server/common/authority/JwtTokenProvider.java:
```diff
@@ -0,0 +1,152 @@
+import java.util.stream.Collectors;
+
+
+@Slf4j
```
존재이유는 ?

sin-hyunjin replied:
```java
 public boolean validateToken(String token) {
        try {
            Jwts.parserBuilder().setSigningKey(key).build().parseClaimsJws(token);
            return true;
        } catch (io.jsonwebtoken.security.SecurityException | MalformedJwtException e) {
            log.info("Invalid JWT Token", e);
        } catch (ExpiredJwtException e) {
            log.info("Expired JWT Token", e);
        } catch (UnsupportedJwtException e) {
            log.info("Unsupported JWT Token", e);
        } catch (IllegalArgumentException e) {
            log.info("JWT claims string is empty.", e);
        }
        return false;
    }

```
log를 기록할때 디버깅이나 예외추적 할 때 좋다고 해서 이 어노테이션을 활용해서    
 log로 JWT로 검증 할려고사용 했습ㄴ디ㅏㅇ

doyou1 commented on src/main/java/com/chukapoka/server/user/sevice/UserService.java:
```diff
@@ -85,20 +122,93 @@ public Long signIn(String email, String password) {
+        Optional<User> optionalUser = userRepository.findById(userId);
+
+        if (optionalUser.isPresent()) {
+            // 2. Access Token에서 user ID를 기반으로 Refresh Token 값 가져오기
+            Token refreshTokenEntity = tokenRepository.findByKey(String.valueOf(userId))
+                    .orElseThrow(() -> new RuntimeException("로그아웃된 사용자입니다."));
```
Good 🚀  이 코드처럼 예상대로 처리되지 않았을 때는 Exception 발생시키도록 코드 수정하자 !

doyou1 commented on src/main/java/com/chukapoka/server/user/sevice/UserService.java:
로그인은 로그인 -> 디비 확인 -> 확인 성공시 토큰 생성 및 리턴 (실패시 exception 발생)
회원가입은 회원가입 -> 디비에 사용자정보 추가(추가 실패시 exception 발생) -> 로그인 처리 (디비확인 -> 확인 성공시 토큰 생성 및 리턴(실패시 exception 발생))

doyou1 reviewed (changes requested):
발급된 토큰 집어넣어야만 실행가능한거 확인함 굿굿. 중복 로그아웃을 exception 처리한 것도 좋았음.(그냥 코드 복붙이지 ?)

의도하지 않은 결과일 경우엔 exception 발생시키도록 하자

sin-hyunjin committed:
리팩토링 수정중

doyou1 committed:
- STR 문법쓰지마
- access token 만료, refresh token 유효일 경우, 재발급 기능 추가(/api/user/reissue)
- refresh token claims에 USER_KEY, AUTHORITIES_KEY추가
- 권한 필요 여부에 따른 엔드포인트 정리
    - 누구나
        - "/api/user/emailcheck", "/api/user", "/api/user/authNumber"
    - token 및 권한 필요
        - "/api/user/logout", "api/user/reissue"
- TokenResponseDto 추가
    - 클라이언트에는 access token만 제공
- TokenRequestDto 삭제
    - request body로 token값이 오는 경우는 없다
- endpoint = controller method name = service method name
- UserRepository findByEmailAndEmailType 추가
- TokenRepository findByAtValue 추가
- UserRequestDto에 emailType, actionType(type에서 변경) 추가

doyou1 committed:
- `application.yaml` 수정(postgres -> h2)

doyou1 committed:
수정2
